File: 1766.tree-of-coprimes.py
#
# @lc app=leetcode id=1766 lang=python3
#
# [1766] Tree of Coprimes
#

# @lc code=start
from typing import List
import math


class Solution:
    '''
    1 <= nums[i] <= 50
    So we can first find all the possible coprime pairs from 1 to 50.
    Then build the graph
    From root, we DFS out and keep a dictionary of how many steps each node on path is away from the root. 
    i.e. dict[nums[i]] = [node index, steps]
    we only care about the last appearance of coprimes, and we scan all the possible coprimes and find the minimum index.
    '''
    def getCoprimes(self, nums: List[int], edges: List[List[int]]) -> List[int]:
        coPrimes = [set() for i in range(51)]
        for i in range(1, 51, 1):
            for j in range(i, 51, 1):
                if math.gcd(i, j) == 1:
                    coPrimes[i].add(j)
                    coPrimes[j].add(i)

        graph = [set() for i in range(len(nums))]
        for a, b in edges:
            graph[a].add(b)
            graph[b].add(a)
        
        result = [-1]*len(nums)
        numseen = {}
        onpath = set()

        def dfs(node, steps):
            lastIndex = -1
            minDist = float('inf')
            for cop in coPrimes[nums[node]]:
                if cop in numseen and len(numseen[cop]) > 0 and (steps - numseen[cop][-1][-1]) < minDist:
                    minDist = steps - numseen[cop][-1][-1]
                    lastIndex = numseen[cop][-1][0]
            result[node] = lastIndex
            onpath.add(node)
            if (nums[node] not in numseen):
                numseen[nums[node]] = []
            numseen[nums[node]].append([node, steps])
            for nxt in graph[node]:
                if nxt not in onpath:
                    dfs(nxt, steps + 1)
            onpath.remove(node)
            numseen[nums[node]].pop()
        
        dfs(0, 0)
        return result
                

# @lc code=end

File: 1766.tree-of-coprimes.cpp
/*
 * @lc app=leetcode id=1766 lang=cpp
 *
 * [1766] Tree of Coprimes
 */

// @lc code=start
#include<vector>
#include<unordered_map>
#include<unordered_set>
#include<numeric>
using std::gcd;
using std::unordered_map;
using std::unordered_set;
using std::vector;
class Solution {
private:
    unordered_map<int, vector<vector<int>>> numSeen{};
    unordered_set<int> onPath{};
    vector<unordered_set<int>> coPrimes;
    vector<unordered_set<int>> graph;
    vector<int> Nums;
    vector<int> result;
    void DFS(int node, int steps)
    {
        int lastIndex = -1;
        int minDist = INT_MAX;
        for (int cop : coPrimes[Nums[node]])
        {
            if (numSeen.find(cop) != numSeen.end() && numSeen[cop].size() > 0 && steps - numSeen[cop].back().back() < minDist)
            {
                minDist = steps - numSeen[cop].back().back();
                lastIndex = numSeen[cop].back().front();
            }
        }

        result[node] = lastIndex;
        onPath.insert(node);

        if (numSeen.find(Nums[node]) == numSeen.end())
        {
            numSeen[Nums[node]] = vector<vector<int>>{};
        }

        numSeen[Nums[node]].push_back(vector<int> { node, steps });

        for (int nxt : graph[node])
        {
            if (onPath.find(nxt) == onPath.end())
            {
                DFS(nxt, steps + 1);
            }
        }

        onPath.erase(node);
        numSeen[Nums[node]].pop_back();
    }
public:
    vector<int> getCoprimes(vector<int>& nums, vector<vector<int>>& edges) {
        Nums = nums;
        coPrimes.resize(51, unordered_set<int>{});
        for (int i = 1; i <= 50; i++)
        {
            for (int j = i; j <= 50; j++)
            {
                if (gcd(i, j) == 1)
                {
                    coPrimes[i].insert(j);
                    coPrimes[j].insert(i);
                }
            }
        }

        graph.resize(nums.size(), unordered_set<int>{});
        for (auto edge : edges)
        {
            int a = edge[0];
            int b = edge[1];
            graph[a].insert(b);
            graph[b].insert(a);
        }

        result.resize(nums.size(), -1);
        DFS(0, 0);
        return result;
    }
};
// @lc code=end

